Feature/round1.1

Iol-lshh opened the pull request:
## 📌 Summary
<!--
    어떤 기능/이슈를 해결했는지 요약해주세요.
-->

기능 구현

- User - 회원가입, 조회
- Point - 충전, 조회

## 💬 Review Points
<!--
    리뷰어가 더 효과적인 리뷰를 할 수 있도록 도와주는 부분입니다.
    (1) 리뷰어가 중점적으로 봐줬으면 하는 부분
    (2) 고민했던 설계 포인트나 로직
    (3) 리뷰어가 확인해줬으면 하는 테스트 케이스나 예외 상황
    (4) 기타 리뷰어가 참고해야 할 사항
-->

### Point를 VO로 설계

- 고민: VO로 설계하는게 맞나? 동일성(id)을 갖을 필요에 대한 고민과 확장성에 애로 사항이 생길 우려에 대한 문제
- [feat: 포인트 모델 구현](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/1ca7955200f6c346699836e9f8a1cc6f4f411f8b)
- [feat: user 모델 구현](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/320ad8a1fbbcd0b944e929ca602183b9223e9a5d)

### Point API 탑다운 User와의 경계

- 고민: 어느 고수준까지 point를 탑다운 관점에서 분리해야할까하는 고민
  - point는 user의 하위 도메인 모델로 작성하였기 때문에
- [feat: 포인트 충전, 조회 기능 구현](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/5e5718146eb8a05a17c57ca979ed181d53f1f05d)
- [feat: 포인트 충전, 조회 API 구현](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/a5ee7079aaee1593ee304d11957d23078b671d4b)

```shell
interface
 ├─ point
 └─ user

application
 ├─ point
 └─ user

domain
 └─ user
      ├─ UserEntity
      └─ UserPointVo
```

### Criteria와 JpaSpecification

- 고민: 도메인에서의 Specification (DDD 에릭 에반스)와의 접점. (cf. queryDSL)
- [feat: user 회원가입, 조회 기능 구현](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/fd3aa2f144b32bfc870cc3e05c388101a0e888fa)
- [fix: UserCriteria 구조 내부 구조를 null에 조금 더 안정적이도록 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/8d61dc1ec55415efef9e42e6e4d2dac64a4c2037)

```shell
domain
 └─ user
      ├─ UserEntity
      ├─ UserCriteria
      └─ UserRepository (interface)

infrastructure
 └─ user
      ├─ UserRepositoryImpl (implements UserRepository)
      └─ UserJpaSpecification (UserCriteria → JPA Specification)
```


## ✅ Checklist
<!--
    해당 작업이 완료되었는지 확인하기 위한 체크리스트입니다.
    리뷰어가 확인해야 할 사항을 포함합니다.
    계획중이나 아직 완료되지 않은 작업 또한 `TODO -` 로 작성해주세요.  

    ex.
    - [ ] 테스트 코드 포함
    - [ ] 불필요한 코드 제거
    - [ ] README or 주석 보강 (필요 시)
-->

- [v]  테스트는 모두 통과해야 하며, `@Test` 기반으로 명시적으로 작성
- [v]  각 테스트는 테스트 명/설명/입력/예상 결과가 분명해야 함
- [v]  E2E 테스트는 실제 HTTP 요청을 통해 흐름을 검증할 것

Iol-lshh committed:
feat: user 모델 구현

Iol-lshh committed:
feat: user 회원가입, 조회 기능 구현

Iol-lshh committed:
feat: user 회원가입, 조회 API 구현

Iol-lshh committed:
feat: 포인트 모델 구현

Iol-lshh committed:
feat: 포인트 충전, 조회 기능 구현

Iol-lshh committed:
feat: 포인트 충전, 조회 API 구현

Iol-lshh committed:
fix: UserCriteria 구조 내부 구조를 null에 조금 더 안정적이도록 수정

Iol-lshh committed:
fix: 명칭 수정 및 @MockitoSpyBean 이용으로 반영

Iol-lshh committed:
fix: 모델 테스트 케이스 추가 및 수정

Iol-lshh committed:
fix: Gender를 UserEntity 안으로 이동

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserPointVo.java:
```diff
@@ -0,0 +1,34 @@
+import lombok.Getter;
+
+@Getter
+@Embeddable
+@Table(name = "point")
```
`@Embeddable` + `@Table` 이면 Embeddable 만 적용되지 않나요? 🤔

Iol-lshh replied:
넵 맞습니다.
엔티티로 작성했던 PointEntity를 UserPointVo로 수정하면서 지나쳤던것같습니다..

- [fix: point 모델과 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/7bf7dbb877574844bbc5338187e3152de49bd436)

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserPointVo.java:
```diff
@@ -0,0 +1,34 @@
+        }
+        var newOne = init();
+        newOne.amount = this.amount + amount;
+        return newOne;
```
요렇게 리팩토링하는게 낫지 않을까요? 불변성을 지키기 위해 VO 로 선정한 것 같은데, 생성 후에 직접 필드 수정하고 있기도 하고.. 안티패턴일 수 있을 것 같아요 😄 
```java
return new UserPointVo(this.amount + amount);
```

Iol-lshh replied:
- [fix: point 모델과 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/7bf7dbb877574844bbc5338187e3152de49bd436)
- [fix: User 모델 수정 및 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/4e088b9537ed1327ffa99624e9ad4b5803193e62)

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserPointVo.java:
```diff
@@ -0,0 +1,34 @@
+@Table(name = "point")
+public class UserPointVo {
+
+    private Long amount;
```
`@Column(nullable = false)` 안 달아줘도 될까용?

Iol-lshh replied:
final을 붙일 수 없으니 `@Column(nullable = false)`을 달면 되겠군요!

- [fix: point 모델과 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/7bf7dbb877574844bbc5338187e3152de49bd436)

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserEntity.java:
```diff
@@ -0,0 +1,78 @@
+    public UserPointVo charge(Long amount) {
+        this.point = this.point.charge(amount);
+        return this.point;
+    }
```
return 값 가질 필요는 없지 않을까요? ㅎㅎ

Iol-lshh replied:
- [fix: User 모델 수정 및 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/4e088b9537ed1327ffa99624e9ad4b5803193e62)

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserCriteria.java:
```diff
@@ -0,0 +1,27 @@
+@Getter
+public class UserCriteria {
+    private final List<Criterion> criteria;
+
```
추상화를 고려한 작성으로 볼 수 있는데, record 로 작성하는 건 어떨까요?
또한 열려있는게, 특정 Usecase (`byLoginId`) 밖에 없어 불필요한 추상화로 보이기도 하네요 🤔

Iol-lshh replied:
UserCriteria 생성시, 반드시 제공하는 정적 팩토리 메서드 패턴만 사용하려는 의도였는데요.
생각해보니 list의 제네릭이 Criterion이기 때문에, 안전하겠다는 생각으로 record로 수정했습니다.

특정 Usecase (byLoginId) 밖에 없어 불필요한 추상화 => 처음 Criteria 객체를 접해서, 시도해봤습니다.
많이 과해보이나요..?

- [fix: User Criteria 레코드로 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/64eb7e2954d0f97fd9005b13c4ce84b9a1324daa)

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserCommand.java:
```diff
@@ -0,0 +1,17 @@
+package com.loopers.domain.user;
+
+public interface UserCommand {
```
별다른 추상명세를 두는게 아니라 inner class 목적인 것 같아서, class 여도 충분하지 않을까 싶은데 어떤가요?

Iol-lshh replied:
```java
record Create (
        String loginId,
        UserEntity.Gender gender,
        String birthDate,
        String email
) implements UserCommand
```
UserCommand 컨테이너 안의 객체들의 타입에도 UserCommand를 구현한다고 표현해주는게 어떨까 생각했었는데요.

1. 구현 의미를 추가
2. 이미 컨테이너 안의 객체이므로 굳이 interface를 구현하는 구조는 과하다.

2번이 더 좋은 지향점이라고 보시는군요..!

- [fix: UserCommand 컨테이너 역할만 하도록 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/91d55adaaa1941642e1c6c99d6875391b9a4f904)

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/application/point/PointFacade.java:
```diff
@@ -0,0 +1,36 @@
+        if(targetUser.isEmpty()) {
+            return null;
+        }
+        return PointInfo.from(targetUser.get());
```
`.map().orElse()` 와 같이 체이닝을 이용하는 건 어때요?

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/interfaces/api/ApiControllerAdvice.java:
```diff
@@ -46,6 +47,14 @@ public ResponseEntity<ApiResponse<?>> handleBadRequest(MissingServletRequestPara
         return failureResponse(ErrorType.BAD_REQUEST, message);
     }
 
+    @ExceptionHandler
+    public ResponseEntity<ApiResponse<?>> handleBadRequest(MissingRequestHeaderException e) {
```
👍

hubtwork commented on apps/commerce-api/src/main/java/com/loopers/interfaces/api/user/UserV1Controller.java:
```diff
@@ -0,0 +1,42 @@
+        UserInfo info = userFacade.get(loginId);
+        if (info == null) {
+            throw new CoreException(ErrorType.NOT_FOUND, "사용자를 찾을 수 없습니다: " + loginId);
+        }
```
이건 service 에서 null 반환하고 통테로 검증.. facade 에서 터뜨려라... 를 권하고 싶었던 건데 제 전달 오류입니다 😭
실무에선 컨트롤러에서 제어하지 않는 방향이 더 좋습니당 ㅎㅎ

Iol-lshh replied:
이 지점에서 PointService를 만들지 않는데, 
어떻게 과제 요구사항을 달성하지? 라는 고민때문에, 일단 어떻게든 구현했습니다 ㅎㅎ

facade와 테스트 내용을 예외 체크로 수정합니다.

- [fix: point facade 수정 및 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/aed7ea1c6fa6f680e43b12964e451a485f1370dc)

hubtwork commented on apps/commerce-api/src/test/java/com/loopers/application/point/PointUsecaseIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+    class Charge {
+
+        @DisplayName("존재하지 않는 유저 ID 로 충전을 시도한 경우, 실패한다.")
+        @Test
```
포인트 충전시, 실제로 조회했을 때 충전된 결과가 조회가능하다 라는 테스트가 추가되면 좋을 것 같아요

Iol-lshh replied:
- [fix: User 모델 수정 및 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/4e088b9537ed1327ffa99624e9ad4b5803193e62)
- [fix: point facade 수정 및 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/aed7ea1c6fa6f680e43b12964e451a485f1370dc)

hubtwork commented on apps/commerce-api/src/test/java/com/loopers/domain/user/UserModelTest.java:
```diff
@@ -0,0 +1,97 @@
+    class Create {
+        @DisplayName("ID 가 영문 및 숫자 10자 이내 형식에 맞지 않으면, User 객체 생성에 실패한다.")
+        @ParameterizedTest
+        @ValueSource(strings = {
+                "",
+                "abcdefghijk", "12345678901",
+                " test", "test ", "te st",
+                "테스트", "テスト", "测试",
+                "abc_def", "abc.def", "abc-def",
+        })
```
다양한 케이스를 검증해 주어서 좋네요 🚀

ksk0605 replied:
22👍

hubtwork commented on apps/commerce-api/src/test/java/com/loopers/domain/user/UserModelTest.java:
```diff
@@ -0,0 +1,97 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+
+class UserModelTest {
```
성별 관련된 테스트가 확충되면 좋을 것 같아요 😄

Iol-lshh replied:
- [fix: User 모델 수정 및 테스트 수정](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/4e088b9537ed1327ffa99624e9ad4b5803193e62)

hubtwork commented on apps/commerce-api/src/test/java/com/loopers/interfaces/api/UserV1ApiE2ETest.java:
```diff
@@ -0,0 +1,169 @@
+
+            System.out.println("Response Status: " + response.getStatusCode());
+            System.out.println("Response Body: " + response.getBody());
+
```
불필요한 println 👿

Iol-lshh replied:
- [fix: 시스템 출력 호출 삭제](https://github.com/Iol-lshh/loopers_be_practice_byjava/pull/4/commits/43a62094e98830ef6fa5444150ae643df7154405)

Iol-lshh committed:
fix: point 모델과 테스트 수정
- @Table 삭제
- 반드시 있어야하는 컬럼에 `@Column(nullable = false)` 반영 (+ VO인데 JPA 데이터 모델이므로)
- 포인트 정상 충전 검증 테스트 추가

Iol-lshh committed:
fix: point facade 수정 및 테스트 수정
- facade에서 포인트 조회시, null 대신 예외
- 포인트 충전 반영 테스트 케이스 추가

Iol-lshh committed:
fix: User 모델 수정 및 테스트 수정
- 생성 - 성별 검증 테스트 추가
- 테스트 로직에 생성자 대신, 실제 사용되는 정적 팩토리 메서드로 테스트 수정
- UserEntity charge 반환값 제거
-

Iol-lshh committed:
fix: User Criteria 레코드로 수정

Iol-lshh committed:
fix: User 생성, 조회 기능의 null 반환 수정
- null 반환을 Optional로 수정
- facade에서 null 반환 시점에, null 반환 대신 예외를 발생으로 수정
- 테스트에 이와 같은 내용을 반영하도록 수정

Iol-lshh committed:
fix: 시스템 출력 호출 삭제

Iol-lshh committed:
fix: UserCommand 컨테이너 역할만 하도록 수정

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/application/point/PointFacade.java:
```diff
@@ -0,0 +1,36 @@
+public class PointFacade {
+    private final UserService userService;
+
+    @Transactional
+    public PointInfo get(String loginId) {
+        UserCriteria criteria = UserCriteria.byLoginId(loginId);
+        var targetUser = userService.find(criteria);
+        if (targetUser.isEmpty()) {
+            throw new CoreException(ErrorType.NOT_FOUND, "사용자를 찾을 수 없습니다: " + loginId);
+        }
+        return PointInfo.from(targetUser.get());
+    }
```
[단순 질문]
조회용 함수에는 readOnly=true 설정을 많이들 사용하잖아요! 혹시 성혁님이 사용하지 않으신 이유가 있을까요??

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserPointVo.java:
```diff
@@ -0,0 +1,36 @@
+
+@Getter
+@Embeddable
+public class UserPointVo {
+
+    @Column(nullable = false)
```
[질문]
성혁님은 고민해보신 영역일 것 같아서 성혁님의 경험과 고민을 듣고싶어서 질문 남깁니다! 

혹시 성혁님은 VO와 Entity 의 차이가 무엇이라고 생각하시나요?  
추가로 DTO까지 포함해서 다룬다면 재미있는 논의가 될 것 같네요 ㅎㅎ

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserEntity.java:
```diff
@@ -0,0 +1,77 @@
+            this.value = value;
+        }
+
+        public static Gender from(String value) {
+            if (value == null || value.isBlank()) {
+                throw new CoreException(ErrorType.BAD_REQUEST, "성별은 비어있을 수 없습니다: " + value);
+            }
+
+            return switch (value) {
+                case "남" -> MALE;
+                case "여" -> FEMALE;
+                default -> throw new CoreException(ErrorType.BAD_REQUEST, "지원하지 않는 성별입니다: " + value);
+            };
+        }
```
[질문] 

```java
public static Gender from(String value) {
    if (value == null || value.isBlank()) {
        throw new CoreException(ErrorType.BAD_REQUEST, "성별은 비어있을 수 없습니다: " + value);
    }

    return Arrays.stream(Gender.values())
            .filter(g -> g.korean.equals(value))
            .findFirst()
            .orElseThrow(() ->
                new CoreException(ErrorType.BAD_REQUEST, "지원하지 않는 성별입니다: " + value));
}
```
저같은 경우에는 Enum을 다룰때 각 요소들 중 일치하는 걸 찾을 때에는 스트림을 사용하는 것을 선호합니다! 이렇게 되면 혹시 추가나 변경, value등의 값이 변경되더라도 기존 코드를 수정할 필요가 없기 때문이에요. 

아마 성혁님도 고려하신 부분일텐데 (물론 성별이 변경 가능성이 높은 도메인은 아니지만)
Enum의 요소들을 찾을 때 switch case 문으로 검사하는 것을 선택하신 이유가 있는지 궁금합니다!

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserPointVo.java:
```diff
@@ -0,0 +1,36 @@
+        return new UserPointVo(0L);
+    }
+
+    public static UserPointVo plus(UserPointVo before, Long amount) {
+        if (amount == null || amount == 0) {
+            throw new CoreException(ErrorType.BAD_REQUEST, "포인트 충전 금액은 0이거나 null일 수 없습니다: " + amount);
+        }
+        if (amount < 0) {
+            throw new CoreException(ErrorType.BAD_REQUEST, "포인트 충전 금액은 음수일 수 없습니다: " + amount);
+        }
+
+        return new UserPointVo(amount + before.amount);
```
[질문]
혹시 해당 함수를 static으로 설정하신 이유가 있나요?

ksk0605 replied:
[제안]
0과 같은 매직넘버도 분리하는 건 어떻게 생각하시나요!

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserService.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Transactional
+    public UserEntity create(UserCommand.Create command) {
+        var byLoginId = UserCriteria.byLoginId(command.loginId());
```
[질문]
타입추론을 선호하시는군요! 저도 var 좋아합니다 ㅎㅎ

제가 좋아하는 이유는 변수의 `= 오른쪽`의 결과인 **타입**이 아니라 **행위**에 집중하는 것이 더 좋은 코드를 만드는 관점이라고
생각하기 때문입니다. 

성혁님은 어떤 이유로 선호하시나요?

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserService.java:
```diff
@@ -0,0 +1,32 @@
+    @Transactional
+    public UserEntity create(UserCommand.Create command) {
+        var byLoginId = UserCriteria.byLoginId(command.loginId());
+        if (userRepository.exists(byLoginId)) {
+            throw new CoreException(ErrorType.CONFLICT, "이미 존재하는 로그인 ID입니다.");
+        }
```
[질문]
이 검증 부분의 위치도 고민을 하셨을 것 같아요. 
경계를 Service에 두신 이유가 무엇인가요?

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/domain/user/UserValidator.java:
```diff
@@ -0,0 +1,47 @@
+import java.time.DateTimeException;
+import java.time.LocalDate;
+
+public class UserValidator {
+    public static void validateLoginId(String loginId) {
+        if (loginId == null || loginId.isBlank()) {
+            throw new CoreException(ErrorType.BAD_REQUEST, "아이디는 비어있을 수 없습니다: " + loginId);
+        }
+        if (!loginId.matches("^[a-zA-Z0-9]+$")) {
+            throw new CoreException(ErrorType.BAD_REQUEST, "아이디는 영문 및 숫자만 허용됩니다: " + loginId);
+        }
+        if (loginId.length() > 10) {
+            throw new CoreException(ErrorType.BAD_REQUEST, "아이디는 10자 이하여야 합니다: " + loginId);
+        }
+    }
+
```
[단순 질문]
Validator 를 따로 구현하신 점 인상깊네요! 
무엇을 검증해야하는지 해당 클래스 파일만 봐도 쉽게 도메인 요구사항을 파악할 수 있을 것 같아요. 

단순한 궁금증입니다만, 저는 Validator 객체는 스프링 빈에서 관리하는 컴포넌트로 등록하는 것을 선호합니다. 
이유는 Validator 또한 우리의 비즈니스를 드러내는 객체이기 때문에 Service 처럼 일종의 도메인 객체라고 보기 때문입니다. 

그래서 따로 검증용 객체를 분리한다면 static으로 선언하기 보다는 @Component를 달아서 사용하는 편인데요. 
static 함수를 선호하시는 이유가 있다면 무엇인지 궁금합니다!

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/infrastructure/user/UserJpaSpecification.java:
```diff
@@ -0,0 +1,25 @@
+import jakarta.persistence.criteria.Predicate;
+import org.springframework.data.jpa.domain.Specification;
+
+public class UserJpaSpecification {
+    public static Specification<UserEntity> with(UserCriteria criteria) {
+        return (root, query, criteriaBuilder) -> {
+
+            Predicate predicate = criteriaBuilder.conjunction();
+            for(var criterion: criteria.criteria()) {
+                if (criterion instanceof UserCriteria.ByLoginId(String loginId)) {
+                    predicate = criteriaBuilder.and(predicate,
+                            criteriaBuilder.equal(root.get("loginId"), loginId));
+                    break;
+                }
+            }
+            // Add more criteria handling as needed
+
+            return predicate;
+        };
+    }
```
[감탄]
이거 너무 재밌어요. 저도 나중에 활용해봐야겠습니다 ㅎㅎ

ksk0605 commented on apps/commerce-api/src/main/java/com/loopers/infrastructure/user/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Override
+    public Optional<UserEntity> find(UserCriteria criteria) {
+        var spec = UserJpaSpecification.with(criteria);
+        var list = jpaRepository.findAll(spec);
+        if (list.size() > 1) {
+            throw new CoreException(ErrorType.INTERNAL_ERROR, "More than one user found");
+        }
+        return list.isEmpty() ? Optional.empty() : Optional.of(list.getFirst());
+    }
```
[질문]
이 함수만 에러메시지가 영어군요! 이유가 있을까요?

ksk0605 commented on apps/commerce-api/src/test/java/com/loopers/application/user/UserUsecaseIntegrationTest.java:
```diff
@@ -0,0 +1,125 @@
+            UserInfo result = userFacade.get("testUser");
+
+            // assert
+            assertNotNull(result);
+            assertNotNull(result.id());
+            assertEquals(expected, result);
```
[질문]
다른 코드를 보니 AssertJ 라이브러리를 사용하는 테스트(assertThat(result))도 있고 이것처럼 Junit(assertXXX(result))를 그대로 사용하는 테스트도 있네요. 

둘 중 어느것을 사용해야할 때를 구분하는 기준이 있으신가요?

ksk0605 reviewed (commented):
역시 성혁님 코드는 👍
굉장히 꼼꼼하고 고민이 많이 드러나는 코드 잘 봤습니다. 
[이런 식으로] 리뷰 상단에 어떤 성격의 코멘트인지 남겨두었습니다만, 
그냥 전부 경험이 적은 한 사람의 소견 정도로 봐주시면 감사하겠습니다 .. ㅋㅋ

많이 배웠습니다. 고생하셨습니다!
